### ISSUE TITLE
Very poor performance from trees?

### ISSUE THREAD

**Issue by deklanw** [online-ml/river#302]
I've been trying out the different classifiers available in creme for multiclass classification. The best two I've found are:

```
# BernoulliNB
# [1,000] Accuracy: 90.49% – 0:00:02 – 125.8KiB
# [10,000] Accuracy: 95.97% – 0:00:25 – 151.3KiB
# [29,000] Accuracy: 96.01% – 0:01:11 – 161.3KiB

# LogisticRegression
# [1,000] Accuracy: 78.08% – 0:00:00 – 124.8KiB
# [10,000] Accuracy: 91.82% – 0:00:07 – 130.5KiB
# [29,000] Accuracy: 93.48% – 0:00:19 – 131.5KiB
```

I've noticed that in particular all of the tree implementations perform poorly

```
# RandomForest
# [1,000] Accuracy: 18.32% – 0:00:06 – 10.1MiB
# [10,000] Accuracy: 19.39% – 0:01:14 – 12.4MiB
# [29,000] Accuracy: 28.55% – 0:03:41 – 27.5MiB

# RandomForest n_trees = 200
# [1,000] Accuracy: 6.81% – 0:01:25 – 199.7MiB
# [10,000] Accuracy: 16.29% – 0:15:07 – 246.2MiB
# [29,000] Accuracy: 27.81% – 0:49:27 – 544.3MiB

# AdaBoost DecisionTree
# [1,000] Accuracy: 19.62% – 0:00:05 – 9.8MiB
# [10,000] Accuracy: 19.96% – 0:01:07 – 12.4MiB
# [29,000] Accuracy: 27.84% – 0:04:35 – 25.3MiB

# DecisionTree
# [1,000] Accuracy: 3.10% – 0:00:00 – 1.1MiB
# [10,000] Accuracy: 4.14% – 0:00:10 – 1.3MiB
# [29,000] Accuracy: 20.31% – 0:00:29 – 3.0MiB
```

Sidenote:
The RandomForest with n_trees=200 gets up to 544MB after only 29k samples! If the number of trees is fixed why is the memory usage growing so much in this case?

I'm using default settings unless otherwise noted. Why are these trees performing so poorly? I went to check the performance of a vanilla batch RandomForest from sklearn with these 29k samples and a test size of 0.3. It got 96% accuracy.

**Comment by MaxHalford** [comment]
Hey there. 

> The RandomForest with n_trees=200 gets up to 544MB after only 29k samples! If the number of trees is fixed why is the memory usage growing so much in this case?

Not sure what you mean. The decision tree is consuming `3.0MiB`, and the random forest is consuming `544.3MiB` with 200 decision trees. That seems reasonable, right?

> I'm using default settings unless otherwise noted. Why are these trees performing so poorly? 

Alas at the moment the decision tree is very sensitive to hyper parameters. I could tell what hyperparameters if I could take a look at your dataset.

**Comment by deklanw** [comment]
Thanks for the quick response

> That seems reasonable, right?

I guess I should have asked: why is the memory usage for decision trees increasingly linearly? They have a finite number of parameters, right?

> I could tell what hyperparameters if I could take a look at your dataset.

I saw your presentation where you did a League of Legends ML thing. Funnily enough that's also my situation. I'm detecting League roles based on various features I'm extracting from the match/timeline. I'm not sure how to best summarize my data for you. I'll give you a sense:

I have categorical features: champion id, keystone id, rune subtree id, summoner combination, game rank. There are about 150~ champs, 17~ keystones, 4~ subtrees, 36~ possible summoner spells, 9~ ranks. I also have various numerical features like wards placed, wards killed, cs, jg cs, times seen in top lane, xp, gold, kills, assists, solo kills in top lane, solo deaths in top lane, etc, etc. I'm using the StandardScaler for all of these features.

Sidenote: how does the BernoulliNB binarize these numerical features?

I'm guessing the BernoulliNB its performing well because the champs gets OneHot encoded into a bunch of binary variables, and champ is pretty highly predictive of role (as you know), not to mention all the other categorical features.

**Comment by MaxHalford** [comment]
Sorry for some reason I missed the notification of your comment!

> I guess I should have asked: why is the memory usage for decision trees increasingly linearly? They have a finite number of parameters, right?

Well most of the memory is taken by the histograms that represent the distribution of each feature with respect to each class. There certainly is a maximum amount of memory, but it's quite high with the default parameters.

For the performance, it's really hard to help out more without having some data to work with. I'm surprise the performance was that bad, but then again it's still early days for creme. Could you maybe try out [onelearn](https://github.com/onelearn/onelearn) and tell me if it works better? I'm almost certain it will. I've had great results with it.

> Sidenote: how does the BernoulliNB binarize these numerical features?

Basically, anything that isn't 0 is treated as positive :)

**Comment by deklanw** [comment]
Np.

I actually had a bug when I was generating these features. Here is the updated performance

```
BernoulliNB
[10,000] Rolling of size 10000 Accuracy: 99.59% – 0:00:15 – 157.46 KB
[20,000] Rolling of size 10000 Accuracy: 99.77% – 0:00:28 – 163.85 KB
[30,000] Rolling of size 10000 Accuracy: 99.72% – 0:00:44 – 167.8 KB
[40,000] Rolling of size 10000 Accuracy: 99.87% – 0:01:00 – 170.8 KB
[50,000] Rolling of size 10000 Accuracy: 99.79% – 0:01:15 – 173.31 KB
[60,000] Rolling of size 10000 Accuracy: 99.78% – 0:01:31 – 174.98 KB
[70,000] Rolling of size 10000 Accuracy: 99.72% – 0:01:45 – 176.08 KB
[80,000] Rolling of size 10000 Accuracy: 99.83% – 0:02:00 – 177.5 KB
[90,000] Rolling of size 10000 Accuracy: 99.85% – 0:02:14 – 179.82 KB
[100,000] Rolling of size 10000 Accuracy: 99.73% – 0:02:32 – 180.28 KB

OneVsRestClassifier(LogisticRegression)
[10,000] Rolling of size 10000 Accuracy: 99.58% – 0:00:12 – 147.68 KB
[20,000] Rolling of size 10000 Accuracy: 99.85% – 0:00:24 – 148.24 KB
[30,000] Rolling of size 10000 Accuracy: 99.84% – 0:00:41 – 148.51 KB
[40,000] Rolling of size 10000 Accuracy: 99.92% – 0:00:57 – 148.79 KB
[50,000] Rolling of size 10000 Accuracy: 99.82% – 0:01:10 – 149.34 KB
[60,000] Rolling of size 10000 Accuracy: 99.89% – 0:01:23 – 149.34 KB
[70,000] Rolling of size 10000 Accuracy: 99.85% – 0:01:35 – 149.34 KB
[80,000] Rolling of size 10000 Accuracy: 99.88% – 0:01:46 – 149.61 KB
[90,000] Rolling of size 10000 Accuracy: 99.87% – 0:01:59 – 150.16 KB
[100,000] Rolling of size 10000 Accuracy: 99.86% – 0:02:11 – 150.16 KB

DecisionTreeClassifier
[10,000] Rolling of size 10000 Accuracy: 4.16% – 0:00:09 – 1.29 MB
[20,000] Rolling of size 10000 Accuracy: 17.19% – 0:00:19 – 1.5 MB
[30,000] Rolling of size 10000 Accuracy: 30.92% – 0:00:29 – 1.57 MB
[40,000] Rolling of size 10000 Accuracy: 62.81% – 0:00:39 – 2.14 MB
[50,000] Rolling of size 10000 Accuracy: 78.86% – 0:00:48 – 2.36 MB
[60,000] Rolling of size 10000 Accuracy: 87.26% – 0:01:01 – 2.75 MB
[70,000] Rolling of size 10000 Accuracy: 97.20% – 0:01:18 – 2.95 MB
[80,000] Rolling of size 10000 Accuracy: 98.26% – 0:01:32 – 3.07 MB
[90,000] Rolling of size 10000 Accuracy: 98.73% – 0:01:44 – 3.76 MB
[100,000] Rolling of size 10000 Accuracy: 98.72% – 0:01:59 – 4.01 MB

RandomForestClassifier n_trees=50
[10,000] Rolling of size 10000 Accuracy: 16.16% – 0:05:40 – 60.11 MB
[20,000] Rolling of size 10000 Accuracy: 37.00% – 0:13:25 – 68.97 MB
[30,000] Rolling of size 10000 Accuracy: 58.35% – 0:18:24 – 71.83 MB
[40,000] Rolling of size 10000 Accuracy: 87.83% – 0:24:34 – 98.41 MB
[50,000] Rolling of size 10000 Accuracy: 96.03% – 0:30:09 – 111.44 MB
[60,000] Rolling of size 10000 Accuracy: 96.04% – 0:35:39 – 128.96 MB
[70,000] Rolling of size 10000 Accuracy: 96.25% – 0:41:59 – 144.09 MB
[80,000] Rolling of size 10000 Accuracy: 96.53% – 0:48:51 – 152.73 MB
[90,000] Rolling of size 10000 Accuracy: 98.57% – 0:57:02 – 160.96 MB
[100,000] Rolling of size 10000 Accuracy: 99.33% – 1:04:43 – 182.29 MB
```
Much better, but still not great and in particular takes a long time to reach reasonable performance.

AdaptiveRandomForestClassifier from scikit-multiflow, (with 10 trees) reaches 98%+ accuracy very quickly in my testing https://scikit-multiflow.readthedocs.io/en/stable/api/generated/skmultiflow.meta.AdaptiveRandomForestClassifier.html#skmultiflow.meta.AdaptiveRandomForestClassifier

See
![image](https://user-images.githubusercontent.com/11601778/84600226-d94e8680-ae45-11ea-87c0-7c611f6fa4b4.png)

ARF = AdaptiveRandomForest
HT = HoeffdingTree

I hadn't heard of onelearn. There is no equivalent of `progressive_val_score` so I just ran it in a batch way on 950k samples with an 80/20 train/test split. I still used `partial_fit`. Accuracy was `98.7%`

After fixing my features I think I'm going to just go with logistic regression. It seems plenty performant and fast in my case. I do wonder about concept drift though. I see creme doesn't support ADWIN or the like. How will creme's logistic regression perform with concept drift?

**Comment by MaxHalford** [comment]
Have you played with the `patience` parameter? Lowering it "should" speed up convergence but might increase the computational cost :)

**Comment by abhibha1807** [comment]
Hey, 
I am trying to code a POS tagger, on the Conll Dataset [https://www.clips.uantwerpen.be/conll2000/chunking/](url) I am using a DecisionTree Classifier for the same. But am getting a poor F1 score of 0.15 
DecisionTree Classifier 
[1,000] F1: 0.169169
[2,000] F1: 0.152076
[3,000] F1: 0.145382
[4,000] F1: 0.141285
[5,000] F1: 0.142028
[6,000] F1: 0.147358
[7,000] F1: 0.149879
[8,000] F1: 0.149644
[9,000] F1: 0.150906
[10,000] F1: 0.150115
F1: 0.150115
While the sklearn implementation gave me a F1 score of 0.78. Here is my code: [https://github.com/abhibha1807/tcs_internship/blob/master/preprocess_conll.py](url). I am wondering is it because of the skewness of the dataset that the classifier isin't working, as the classifier was only predicting 2-3 labels out of the many pos tags. Any help would be greatly appreciated! :)

**Comment by MaxHalford** [comment]
Hello @abhibha1807, thanks for your comment. I don't have time to look at your code in detail. Could you potentially create a CSV file and send that instead? Kind regards.

**Comment by abhibha1807** [comment]
Hey @MaxHalford, Thanks for reverting back! If am interpreting it right, I have attached the dataset (in csv format) and the log for my code ( I have run the code on the whole dataset ). I also tried fiddling with the hyper parameters but I don't seem to get any improvements. Is there any way in which I can determine the optimal set of hyper parameters for my dataset?. I think I may be erring in that cause as stated above, Decision trees tend to be sensitive to hyperparamaters.  Sorry for bundling you with questions 😅. Thanks in advance :) 
[pos_logs.zip](https://github.com/creme-ml/creme/files/4791663/pos_logs.zip)
